### ISSUE TITLE
Small object under 15px detection

### ISSUE THREAD

**Issue by sinuku** [ultralytics/ultralytics#981]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hi, I am a beginner for yolov8.

I have used the yolox model before, and the results are worse than before for small objects.

Are there any tricks to improve performance by modifying the models.v8.yaml file?

**Comment by JustasBart** [comment]
Hi @sinuku your problem sounds a bit trivial at best and to me it sounds like you would need something perhaps more specialised to achieve what you are describing. The problem is that yolov5 and yolov8 are based on the COCO dataset which has practically no small objects or not that small anyway...

Yeah, I can't really think of an obvious solution to your problem. The only advice I could give is to try and show it some negative data and perhaps turn off the augmentations like so:

hsv_h=0.015 \
hsv_s=0.75 \
hsv_v=0.05 \
degrees=0.0 \
translate=0.1 \
scale=0.05 \
shear=0.0 \
flipud=0.5 \
fliplr=0.5

Other than that you might have to research this deeper, good luck! :rocket:

**Comment by patrickwasp** [comment]
see if you get better results using detectron2. I found it works on small objects much better.

**Comment by majnas** [comment]
You might get better accuracy using SAHI algorithm 
Link: https://github.com/obss/sahi

**Comment by CPFelix** [comment]
Try yolov8-P2.It has effect on small objects.

**Comment by glenn-jocher** [comment]
@CPFelix hello, 

Thanks for reaching out! Yes, YOLOv8-P2 is a variant of the YOLOv8 model designed for better performance on small objects. It includes additional feature pyramid levels compared to the original YOLOv8 model, which enables it to detect and localize smaller objects more accurately.

To use YOLOv8-P2, you need to modify the `models.v8.yaml` file to set the `backbone` parameter to `yolov5m6-p2`. This will load the YOLOv5m6-p2 backbone network instead of the standard YOLOv5 backbone and allow you to use YOLOv8-P2.

I hope this helps you solve the issue you are experiencing! Don't hesitate to let us know if you have any further questions or concerns.

**Comment by sinuku** [comment]
@glenn-jocher Thanks for your help.
Can I just use `model = YOLO("yolov8s-p2.yaml")` and load `yolov8s.pt` for the pretrained model?

**Comment by kurkurzz** [comment]
Hey I did what you said:
```py
model = YOLO('yolov8s-p2.yaml')
model.load_state_dict(torch.load('yolov8s.pt'))
```
but I got `YOLO' object has no attribute 'load_state_dict'`

So I changed it to:
```py
model = YOLO('yolov8s-p2.yaml')
model.model.load_state_dict(torch.load('yolov8s.pt'), strict=False)
```
I got no error but I got incompatible keys warning. Is my implementation correct? If not, can you provide the exact code to achieve this.

**Comment by CPFelix** [comment]
@kurkurzz Use like this:
model = YOLO('yolov8s-p2.yaml').load('yolov8s.pt')

**Comment by alpha-31** [comment]
No this gives an issue of depth_multiple 
<img width="958" alt="v8error" src="https://github.com/ultralytics/ultralytics/assets/60798854/df872727-e09e-47b1-9ab4-58973c9f2ae4">

**Comment by Han6260** [comment]
May I ask if the training dataset is relatively large, can the best. pt file obtained by using yolov8-p2.yaml and loading the weight of yolov8x be used to detect smaller objects using a camera

**Comment by glenn-jocher** [comment]
@Han6260 hello,

Certainly, you can use the `yolov8-p2.yaml` model which has been trained on a relatively large dataset and has imported the yolov8x weights to detect smaller objects using a camera. The 'P2' in `yolov8-p2.yaml` refers to the inclusion of additional feature pyramid levels, thus creating a detection model that can better recognize and accurately localize smaller objects.

Bear in mind that the actual inference performance on small objects might vary from one scenario to another. For instance, the characteristics of the camera, including resolution and the distance from the objects all play a role. Similarly, lighting conditions or the amount of clutter in the scene can also be factors.

In case you face challenges achieving desired detection accuracy, you might want to experiment with different configurations, such as adjusting the anchor box sizes to better match your small objects or tuning other hyperparameters in the `yolov8-p2.yaml` file.

I hope this helps! Should you have any further questions, feel free to ask.

**Comment by Han6260** [comment]
I trained feature maps with larger sizes to attempt to detect categories of smaller targets, but the resulting pt file did not perform well on the camera. But I don't know what the reason is.

**Comment by glenn-jocher** [comment]
Hello @Han6260,

Detecting smaller objects in images can indeed be challenging due to numerous factors. The clarity and resolution of images, as well as the relative size of the objects of interest, greatly impact the model's performance.

In terms of training the model with larger feature maps, remember that while this might help maintain more details (which could potentially improve the detection of small objects), it also increases the complexity of the model and the computational resource requirements.

If the model didn't perform as expected, it's possible that some hyperparameters might need further tuning. Parameters such as learning rate or the architecture's settings (e.g., number of layers, type of layers, etc.) can greatly influence results.

Also, consider the characteristics of your training data. If your training data is vastly different from the real-time data captured by your camera (in terms of image quality, lighting conditions, scale of objects, etc.), the model might struggle to generalize well. In this case, you might want to incorporate more diverse or representative samples into your training set.

Without more specifics, it's difficult to pinpoint the exact reason, but hopefully these general pointers help guide you toward improving your model's performance. Finally, remember that evaluating model performance only based on live camera data can be quite deceptive due to the many variables involved (camera angle, lighting, distance, etc.). Please consider using a validation set for a more robust comparison. 

Let us know if we can assist you further, and good luck with your project!

**Comment by jrsaffy** [comment]
Hi @glenn-jocher,

I performed the step that @CPFelix suggested `model = YOLO('yolov8s-p2.yaml').load('yolov8s.pt')` and I was able to load the model, but it looks like I'm not getting any detections at all.  I've gotten detections when I simply use `model = YOLO()` with the same camera in the same environment.  I'm hoping to detection small objects, at least small enough to match yolov7's performance in that area

Thanks!

**Comment by glenn-jocher** [comment]
Hello @jrsaffy,

Thanks for trying out YOLOv8 and I'm sorry to hear that you're having some problems with detection. Your point of comparison, YOLOv7, and the YOLOv8-S P2 model are quite different in terms of architecture and training regime, so it's not necessarily surprising that their behavior varies.

Concerning your specific issue, it could be that the `yolov8s-p2.yaml` architecture combined with the `yolov8s.pt` weights just aren't performing well on your specific data. Different architectures can behave differently with different types of input and your use case of small object detection can be particularly challenging.

Remember, one of the most crucial parts in machine learning tasks is the choice and tuning of the model to match the data and task. It is worth considering different YOLOv8 models and tuning their respective hyperparameters accordingly. 

If you're still encountering difficulties, please don't hesitate to elaborate more on your issue. Include details such as the nature of your images, the small objects you are trying to detect, and any other relevant details to give us a clearer picture of the problem. This will enable us to provide more targeted advice.

**Comment by Han6260** [comment]
I want to test the types of cars under the category. I built a dataset based on the appearance of different car models for training. When training for feature extraction, the appearance, color, pattern, and other information of the car should be extracted. Due to the fact that some cars are too far away from the camera, resulting in smaller detected cars, there may be situations where they are mistaken or unrecognized. I initially used the yolov8n model, but later used the yolov8n-P2 model, but the results were not very good, and there were cases of misidentification or unrecognizability。
Looking forward to your answer, thank you

**Comment by glenn-jocher** [comment]
Hello @Han6260,

Thanks for your question! You're on the right track using the YOLOv8 model for object detection. However, it seems like you are facing challenges in training the model to detect specific car models, especially when some cars are too far from the camera resulting in smaller detected cars.

The issue of misidentification or failing to detect certain car models could be due to several factors. For instance, the dataset variety and quality could be influencing the performance. It is important to ensure that your dataset has a wide range of car models, viewed from different angles, lighting and weather conditions, colors, and at different distances from the camera.


Additionally, the sophistication of the model, in terms of architecture and other settings, could also be playing a role in performance. Switching from the YOLOv8 model to YOLOv8N-P2 is indeed a good idea as the latter introduces a feature pyramid network, which should theoretically increase the model's ability to detect objects at different scales. However, the performance might still be suboptimal if it is not properly optimised for the task at hand. Therefore, it could be worth fine-tuning the model's architecture or settings, based on the results you are observing.


Finally, training and optimising the model carefully is crucial to ensure its robustness and generalisation capabilities. This includes setting appropriate loss function weights, choosing a good optimizer and learning rate, performing data augmentation, and so forth.


I hope these suggestions help inform your next steps! Please don't hesitate to ask if you have further questions or clarifications.

**Comment by hellowqc** [comment]
> 你好@Han6260,
> 
> 谢谢你的提问！使用 YOLOv8 模型进行对象检测，您走在正确的道路上。然而，您似乎在训练模型来检测特定汽车模型方面面临挑战，特别是当某些汽车距离摄像机太远导致检测到的汽车较小时。
> 
> 错误识别或未能检测某些车型的问题可能是由多种因素造成的。例如，数据集的多样性和质量可能会影响性能。重要的是要确保您的数据集具有广泛的汽车模型，从不同的角度、照明和天气条件、颜色以及距相机的不同距离观看。
> 
> 此外，模型在架构和其他设置方面的复杂性也可能对性能产生影响。从 YOLOv8 模型切换到 YOLOv8N-P2 确实是一个好主意，因为后者引入了特征金字塔网络，理论上应该会增加模型检测不同尺度物体的能力。但是，如果没有针对手头的任务进行适当优化，性能可能仍然不是最佳的。因此，可能值得根据您观察到的结果来微调模型的架构或设置。
> 
> 最后，仔细训练和优化模型对于确保其稳健性和泛化能力至关重要。这包括设置适当的损失函数权重、选择良好的优化器和学习率、执行数据增强等等。
> 
> 我希望这些建议对您的下一步行动有所帮助！如果您还有其他问题或需要澄清，请随时询问。

Hi. I'm new in using YOLO.
Can someone please explain to me what is the exact difference between the following three commands?
And when do we use each one?
Load a model
model = YOLO('yolov8n.yaml') # build a new model from YAML
model = YOLO('yolov8n.pt') # load a pretrained model (recommended for training)
model = YOLO('yolov8n.yaml').load('yolov8n.pt') # build from YAML and transfer weights

**Comment by hellowqc** [comment]
> 你好@Han6260,
> 
> 谢谢你的提问！使用 YOLOv8 模型进行物体检测，您走在正确的道路上。然而，您似乎在训练模型来检测特定汽车模型方面面临挑战，特别是当某些汽车距离太空太远导致检测到的汽车缩小时。
> 
> 错误识别或未能检测到某些汽车模型的问题可能是由多个因素引起的。例如，数据集造成的多样性和质量可能会影响性能。重要的是要确保您的数据集具有广泛的汽车模型，从不同的角度、照明和天气条件、颜色以及距焦距的不同距离观看。
> 
> 另外，模型在架构和其他设置方面的复杂性也可能对性能产生影响。从YOLOv8模型切换到YOLOv8N-P2确实是一个好主意，因为最新引入了特征金字塔网络，理论上应该会增加模型检测不同但是，如果没有针对手头的任务进行适当的优化，性能可能仍然不是最佳的。因此，可能值得您根据观察到的结果来调整模型的架构或设置。
> 
> 最后，仔细训练和优化模型以确保其完整性和泛化能力至关重要。这包括设置适当的损失函数权重、选择良好的优化器和学习率、执行数据增强等。
> 
> 我希望这些建议对您的下一步行动有所帮助！如果您还有其他问题或需要澄清，请随时询问。

Hi. I'm new in using YOLO.
Can someone please explain to me what is the exact difference between the following three commands?
And when do we use each one?
model = YOLO('yolov8n.yaml') # build a new model from YAML
model = YOLO('yolov8n.pt') # load a pretrained model (recommended for training)
model = YOLO('yolov8n.yaml').load('yolov8n.pt') # build from YAML and transfer weights

**Comment by Frank-123-123** [comment]
你好   
      我刚接触到yolov8，现在有一些问题，在近距离的检测效果较好，但是在远距离的小目标检测上就是检测不出来 ，请问有什么好的办法或者调整什么样的训练参数适合在小目标检测上

**Comment by glenn-jocher** [comment]
@Frank-123-123 你好呀！👋

远距离小目标检测确实是挑战性的。提升远距离小目标检测效果，一个常见方法是通过调整输入图像的分辨率，尝试增加 `imgsz` 参数值进行训练，使模型能在更高的分辨率下更好地学习细节信息。比如：

```python
model.train(data='your_dataset.yaml', imgsz=640, epochs=100)
```
其中 `imgsz` 可以根据需求和硬件能力适当增加。另外，确保你的数据集中包含足够的远距离小目标样本，这样模型才能学会识别它们。数据增强（如裁剪、缩放）也可能有所帮助。

祝训练顺利！ 🚀

**Comment by haolele123** [comment]
Hello, I would like to ask for advice, we want to achieve a single category detection task of people, and the current input picture is the 1920* collected by the camera* 1080 pictures, the training dataset is the category of people in the coco dataset and some manually labeled pictures, but the model training results are not very good on small targets, so I would like to ask you what good suggestions you have for the input size of the model in this scenario, and at the same time, is there a good way to improve the model of yolov8, I look at the aforementioned yolov8n P2 would be a good idea, but I also thought about whether it would be better to change the default size of the anchor to make the model better adapted to small target tasks. Thank you for your answer

**Comment by glenn-jocher** [comment]
@haolele123 hello! 🌟 For your single-category detection task focused on people, especially concerning small targets in 1920x1080 images, adjusting the input size (`imgsz`) and considering anchor modifications can indeed be helpful strategies.

Given the resolution of your images, you might experiment with increasing the `imgsz` during training to provide the model more detail. For YOLOv8, a starting point might be:

```python
model.train(data='your_data.yaml', imgsz=640, batch=16, epochs=100)
```
If your hardware permits, you could try even larger sizes. Additionally, exploring the P2 variant is a great idea, as it's designed with the capacity to better grasp fine-grained details, potentially enhancing small object detection.

Regarding anchor sizes, adapting them to more closely match the dimensions of the people in your dataset might aid the model. This is a bit advanced and requires analyzing the distribution of your target sizes before adjusting.

Remember, achieving excellent detection on small targets often also demands a rich, diverse training set that includes ample examples of such challenges. 

Hope this nudges you in the right direction! Good luck with your project! 🚀

**Comment by djdll** [comment]
Here I found a similar problem, I would like to ask if v8-p2.yaml can improve small object detection according to what you said. So if my scene only has small objects, and the size of small objects is basically the same, then I think it is redundant to keep the large object branch, here I directly remove or simplify the large object convolution operation is it feasible? Or what are the better suggestions to train the network only for small targets, because I see that the "head" part is to cat different layers to operate. If I only send to the neck of a small target, does the result have a big impact on the whole network?

**Comment by glenn-jocher** [comment]
If your scene primarily involves small objects, using the YOLOv8-P2 model can indeed help. Simplifying or removing the large object branch might be feasible, but it requires careful modification to ensure the model remains balanced. Focusing the network on small targets by adjusting the "head" and "neck" layers could improve performance, but extensive testing is recommended to validate the impact.
